go/libraries/doltcore/table/editor/pk_table_editor.go

```diff
@@ -188,34 +195,42 @@ func (tea *tableEditAccumulator) NewFromCurrent() *tableEditAccumulator {
 		insertedKeys: make(map[hash.Hash]types.Value),
 		addedKeys:    make(map[hash.Hash]types.Value),
 		removedKeys:  make(map[hash.Hash]types.Value),
-		affectedKeys: make(map[hash.Hash]types.Value),
+		affectedKeys: make(map[hash.Hash]*doltKVP),
 	}
 }
 
-// Has returns whether the current tableEditAccumulator contains the given key. This assumes that the given hash is for the given
-// key.
-func (tea *tableEditAccumulator) Has(ctx context.Context, keyHash hash.Hash, key types.Value) (bool, error) {
+// doltKVPIfHas returns a *doltKVP if the current tableEditAccumulator contains the given key, or it exists in the row data.
+// This assumes that the given hash is for the given key.
+func (tea *tableEditAccumulator) doltKVPIfHas(ctx context.Context, keyHash hash.Hash, key types.LesserValuable) (*doltKVP, bool, error) {
 	// No locks as all calls and modifications to tea are done from a lock that the caller handles
-	if _, ok := tea.addedKeys[keyHash]; ok {
-		return true, nil
-	}
-	if _, ok := tea.removedKeys[keyHash]; !ok {
-		// When rowData is updated, prevTea is set to nil. Therefore, if prevTea is non-nil, we use it.
-		if tea.prevTea != nil {
-			pkExists, err := tea.prevTea.Has(ctx, keyHash, key)
-			if err != nil {
-				return false, err
-			}
-			return pkExists, nil
+	if kvp, ok := tea.affectedKeys[keyHash]; ok {
```

Comment:
I think we still need these `addedKeys/removedKeys` checks. If you add then remove then the key isn't really in the `tea`, but will still be logged in `affectedKeys`, so this will return it.

I'd like to see a test confirming that `ON DUPLICATE KEY` isn't triggered when you insert a new, then delete it, then insert it again.

Reply:
So I added this to the bats tests:

```bash
    dolt sql -b -q "INSERT INTO test VALUES ('d','d','d');DELETE FROM test WHERE col_a='d';INSERT INTO test VALUES ('d', '1', '1') ON DUPLICATE KEY UPDATE col_b = '2', col_c='2';"
    run dolt sql -r csv -q "SELECT * from test where col_a = 'd'"
    [ $status -eq 0 ]
    [[ "$output" =~ "d,1,1" ]] || false
```

and it works.  I think there are 2 things here.  First inserts and deletes will not be run as part of the same batch, but I think they could be now and I think this will still work because in the case of a delete we do:

`te.tea.affectedKeys[oldHash] = &doltKVP{k: dOldKeyVal}`

leaving `v` as nil.  Then in this check we do:

```go
	// No locks as all calls and modifications to tea are done from a lock that the caller handles
	if kvp, ok := tea.affectedKeys[keyHash]; ok {
		if kvp.v == nil {
			// if the row has been deleted then return that the row doesn't exist even if it's in the row data
			return nil, false, nil
		} else {
			return kvp, true, nil
		}
	}
```

I've added that comment there to make it clearer as this is easy to overlook.